src/prompt/types.ts

```diff
@@ -1,9 +1,50 @@
 import type { z } from 'zod';
+import type { SetOptional } from 'type-fest';
+import type { Model } from '../index.js';
 
 export namespace Prompt {
   export type ChainParams = Record<string, any> | void;
   export type ChainResult = string | Record<string, any>;
 
+  /**
+   * A runner that iteratively calls the model and handles function calls.
+   */
+  export type Runner<Content extends any = string> = (
+    params: string | Runner.Params,
+    context?: Model.Ctx
+  ) => Promise<Runner.Response<Content>>;
+  export namespace Runner {
+    /** Parameters to execute a runner */
+    export type Params = SetOptional<
+      Model.Chat.Run & Model.Chat.Config,
+      'model'
+    >;
+
+    /** Response from executing a runner */
+    export type Response<Content extends any = string> =
+      | {
+          status: 'success';
+          messages: Prompt.Msg[];
+          content: Content;
+        }
+      | {
+          status: 'error';
+          messages: Prompt.Msg[];
+          error: Error;
+        };
+
+    /** Controls use of functions or tool_calls from OpenAI API */
+    export type Mode = 'tools' | 'functions';
+  }
+
+  /**
+   * A function used to extract data using OpenAI function calling.
+   */
+  export type ExtractFunction<Schema extends z.ZodObject<any>> = (
```

Comment:
one thing I ended up doing for agentic was using normal TS types for generic types and then having any zod schemas be based off those types instead of the other way around which turned out to be cleaner typing-wise because the base types are easier to work with, and the fact that you're using zod is really more of an impl detail.

this is difficult to explain without a before & after example, but it'd allow something like `ExtractFunction`, `AIFunction`, and `Runner`/`Chain` to interop typing-wise more closely since none of their generic types would depend on zod